src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+
+            return node.ReplaceToken(token, token.WithAdditionalAnnotations(RenameAnnotation.Create()));
+        }
+
+        private ForeachInfo GetForeachInfo(ISemanticFactsService semanticFact, SemanticModel model, SyntaxNode foreachStatement, CancellationToken cancellationToken)
+        {
+            var operation = model.GetOperation(foreachStatement, cancellationToken) as IForEachLoopOperation;
+            if (operation == null || operation.Locals.Length != 1)
+            {
+                return null;
+            }
+
+            var foreachVariable = operation.Locals[0];
+            if (foreachVariable == null)
+            {
+                return null;
+            }
+
+            if (!CheckForeachVariable(model, foreachVariable, foreachStatement))
+            {
+                return null;
+            }
+
+            var foreachCollection = RemoveImplicitConversion(operation.Collection);
+            if (foreachCollection == null)
+            {
+                return null;
+            }
+
+            string explicitCast;
+            string countName;
+            GetInterfaceInfo(semanticFact, model, foreachVariable, foreachCollection, out explicitCast, out countName);
+            if (countName == null)
+            {
+                return null;
+            }
+
+            var requireCollectionStatement = CheckRequireCollectionStatement(foreachCollection);
+            return new ForeachInfo(countName, explicitCast, requireCollectionStatement, foreachStatement);
+        }
+
+        private static void GetInterfaceInfo(
+            ISemanticFactsService semanticFact, SemanticModel model, ILocalSymbol foreachVariable, IOperation foreachCollection, out string explicitCast, out string countName)
+        {
+            explicitCast = default;
+            countName = null;
+
+            // go through list of types to find out right set;
+            var foreachType = foreachVariable.Type;
+            if (IsNullOrErrorType(foreachType))
+            {
+                return;
+            }
+
+            // check array case first.
+            var collectionType = foreachCollection.Type;
+            if (IsNullOrErrorType(collectionType))
+            {
+                return;
+            }
+
+            if (collectionType is IArrayTypeSymbol array &&
+                array.Rank == 1 &&
+                semanticFact.IsAssignableTo(array.ElementType, foreachType, model.Compilation))
+            {
+                explicitCast = null;
+                countName = Length;
+                return;
+            }
+
+            // check ImmutableArray case second
+            if (collectionType.OriginalDefinition.Equals(model.Compilation.GetTypeByMetadataName(typeof(ImmutableArray<>).FullName)))
+            {
+                var indexer = GetInterfaceMembers(collectionType, get_Item);
+                if (indexer != null && semanticFact.IsAssignableTo(indexer.ReturnType, foreachType, model.Compilation))
```

Comment:
i feel like things should work even if the types are not assignable.  Indeed, i would always just do the conversions, just always add a cast to the foreachType with a simplification annotation.  Then, if the cast is necessary, it will be kept.  Otherwise it will be removed.

Reply:
it can't determine whether cast is necessary or not. since simplifier only concern the expression. cast is always not necessary for the expression.

cast is needed in case indexer/count is implmeneted as explicit interface. also, cast change what Count name should be used.

also, checking whether value is assignable is needed for case where there are multiple interface with different type is used to find out which type it is converting to.

Reply:
Sorry, not sure I understand.  I'm saying, just generate the following:

```c#
for (var i = 0; i < array.Length; i++) {
    OtherType a = (OtherType)array[i]; // add simplification annotation on this cast.
}
```

This accurately represents what happens in a foreach anyways.  i.e. if i do: 'foreach (OtherType a in b)'

and 'b' does not return 'OtherTypes', then the compiler will insert casts (which can fail at runtime).

So just do the same thing in your translation.  Just insert the cast and have the simplifier remove it if unnecessary.